### ISSUE TITLE
Is there a way to have the installed module source a shell script on load?

### ISSUE THREAD

**Issue by kelseymh** [easybuilders/easybuild-framework#3275]
I know about the `modextravars` and `modextrapaths` settings in EasyConfigs, and they are working nicely for us.  However, for some of our very large packages (Geant4 and ROOT in particular) using these directly would be quite difficult to support.  Those packages provide shell scripts the user is expected to source, which define (in a build-dependent way) a very large number of envvars, including some search paths.

Ideally, either in one of these modules themselves, or at least in my group environment Bundle ("module of modules"), I'd like to have a directive passed through to the .lua file to source the package setup script to do that configuration.

Is this something which is already possible, but is maybe not documented clearly (https://easybuild.readthedocs.io/en/latest/version-specific/easyconfig_parameters.html#modules-parameters)?

**Comment by ocaisa** [comment]
Are you aware Geant4 and ROOT are already supported by EasyBuild? E.g., use `eb --search geant` to find supported versions.

**Comment by kelseymh** [comment]
Yes, we are using them.  However, they do not set all of the G4-specific or ROOT-specific envvars needed for compiling and linking executables against them.  We also see that ROOT doesn't get included in PYTHONPATH as a module, so we cannot get to PyROOT directly (via import ROOT) in a Python job.

But I'm asking the general question, not about a specific module.  If I am writing an .eb file, is there a directive (special variable = value) I can include to have a script sourced at load time?  Or is that too big of an ugly security hole to want to open up?

**Comment by boegel** [comment]
@kelseymh This is not supported in EasyBuild right now, because it's not really supported in the environment modules tools like Lmod either.

I believe the main reason for this is that sourcing a script can't be reversed, implying that `module unload` would not work as expected.

@rtmclay (Lmod developer) and @xdelaruelle (current maintainer of Environment Modules v4): any additional comments on this?

**Comment by xdelaruelle** [comment]
As you say @boegel, the tricky part is reversing the situation once the module gets unloaded.

That is not impossible to do but it would require to track every changes made to the environment by each loaded module to know what to unset when unloading. Which is a major change to implement.

This is not the first time that I hear such use case, so I have it in my *todo* list, but it is not on top of the list.

**Comment by xdelaruelle** [comment]
It seems there is already a *hacky* way to source a shell script in a modulefile.

Let's say we have a basic script shell to source:
```console
$ cat test.sh
export FOO=bar
```

It could be swallowed in a modulefile by using the script provided with Environment Modules, that converts a shell script in a modulefile:  `createmodule.sh`.

```
$ cat testmod
#%Module
eval [exec /usr/bin/env -i /usr/share/Modules/bin/createmodule.sh test.sh]
```

`/usr/bin/env -i` is used here to ensure `createmodule.sh` will provide the same output whether the module is loaded or unloaded

Demonstration:
```console
$ ml ./testmod
$ echo $FOO
bar
$ ml -./testmod
$ echo $FOO

```

It is a *bit* hacky, but it seems to do the job.

I am pretty sure the same thing could be achieved with Lmod, just by replacing the `createmodule.sh` script by its Lmod counterpart: `sh_to_modulefile`.

**Comment by kelseymh** [comment]
@xdelaruelle Thank you!  My GitHub-fu is weak this morning :-(  

I tried searching the [TACC/Lmod](https://github.com/TACC/Lmod) repository, but didn't find "createmodule.sh".  I did [find "sh_to_modulefile"](https://github.com/TACC/Lmod/search?q=sh_to_modulefile&unscoped_q=sh_to_modulefile), but it's not obvious to me how to make use of it.

From reading the [documentation](https://github.com/TACC/Lmod/blob/master/docs/source/260_sh_to_modulefile.rst) that came up in that search, it seems like this would handle even complex "setup" scripts with if blocks (relevant for at least three of my use cases): it runs the script, and effectively "diff"s the new environment vs. the previous.

Is this something an EC write would add to the end of the generated .lua file by way of EB's [modluafooter](https://easybuild.readthedocs.io/en/latest/version-specific/easyconfig_parameters.html#modules-parameters) setting)?

As I noted above, I do realize there are security implications for letting EC writers invoke arbitrary shell scripts at module load time.

**Comment by xdelaruelle** [comment]
I have to amend the proof of concept exposed above as the `env -i` strategy will be too harsh for scripts as they will require at least some environment variable to be set to properly execute. For instance, as soon as the script will try to execute an external command, it will need `PATH` variable to be set to find the external command.

A more suitable strategy would be:
* when loading module:
  - to evaluate script and get corresponding modulefile commands (thanks to `createmodule.sh`)
  - to evaluate those commands in modulefile context
  - to record those commands obtained in an environment variable
* when unloading module:
  - to retrieve the commands saved in environment at load time
  - to evaluate those commands (to get them unset)

New `test.sh` script for demo:

```sh
export FOO=bar
export BAR=$(uname -m)
```

And new proof-of-concept modulefile:

```tcl
#%Module

switch -- [module-info mode] {
    load - display {
        set sourcesh [exec /usr/share/Modules/bin/createmodule.sh test.sh]
        # remove modulefile header
        set sourcesh [string range $sourcesh 13 end] 
        eval $sourcesh
        # record content in environment to remember it when unloading
        setenv TEST_SOURCED_SCRIPT [string map {\n ;} $sourcesh]
    }
    unload {
        if {[info exists env(TEST_SOURCED_SCRIPT)]} {
            eval $env(TEST_SOURCED_SCRIPT)
            # unset variable as we are unloading
            setenv TEST_SOURCED_SCRIPT $env(TEST_SOURCED_SCRIPT)
        }
    }
}
```

Load and unload test of this modulefile:

```console
$ module load ./testmod
$ echo $TEST_SOURCED_SCRIPT 
setenv BAR x86_64;setenv FOO bar
$ echo $FOO
bar
$ echo $BAR
x86_64
$ module unload ./testmod
$ echo $FOO

$ echo $BAR

$ echo $TEST_SOURCED_SCRIPT

$ 
```

It seems doable to extend `module` solutions to have a new modulefile command for that, named for instance `source-sh` with the following usage:
```
source-sh shell script [arg...]
```

**Comment by kelseymh** [comment]
I'm going to close this issue.  @xdelaruelle gave us the clue we needed for this, the `createmodule.sh` utility.

What we did (not for G4) was to run the big nested setup script generated by our "global release" through `createmodule.sh`, creating a "non-EasyBuild" module.  Then, in the EasyConfig where I was wanting to source that script, I just put that new module in as `EXTERNAL_MODULE` (using my just-merged PR #3324 to find it).

The net result is an EasyBuild module for our full software, with the necessary dependencies, and the full environment "script" sourced automatically.
